flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java

```diff
@@ -100,21 +100,37 @@ protected CatalogLoader createCatalogLoader(String name, Map<String, String> pro
   }
 
   @Override
-  public Map<String, String> requiredContext() {
-    Map<String, String> context = Maps.newHashMap();
-    context.put(CatalogDescriptorValidator.CATALOG_TYPE, "iceberg");
-    context.put(CatalogDescriptorValidator.CATALOG_PROPERTY_VERSION, "1");
-    return context;
+  public String factoryIdentifier() {
+    return FlinkCatalogFactoryOptions.IDENTIFIER;
   }
 
   @Override
-  public List<String> supportedProperties() {
-    return ImmutableList.of("*");
+  public Set<ConfigOption<?>> requiredOptions() {
+    final Set<ConfigOption<?>> options = new HashSet<>();
+    options.add(FlinkCatalogFactoryOptions.CATALOG_TYPE);
+    return options;
   }
 
   @Override
-  public Catalog createCatalog(String name, Map<String, String> properties) {
-    return createCatalog(name, properties, clusterHadoopConf());
+  public Set<ConfigOption<?>> optionalOptions() {
+    final Set<ConfigOption<?>> options = new HashSet<>();
+    options.add(FlinkCatalogFactoryOptions.PROPERTY_VERSION);
+    options.add(FlinkCatalogFactoryOptions.URI);
+    options.add(FlinkCatalogFactoryOptions.WAREHOUSE);
+    options.add(FlinkCatalogFactoryOptions.CLIENTS);
+    options.add(FlinkCatalogFactoryOptions.BASE_NAMESPACE);
+    options.add(FlinkCatalogFactoryOptions.HIVE_CONF_DIF);
+    options.add(FlinkCatalogFactoryOptions.CACHE_ENABLED);
+    return options;
+  }
```

Comment:
Do we need to add any of the known additional catalog properties that are used with some of the AWS catalogs?

Also, for catalogs that offer additional options (including custom catalogs), how will users set those? Presently, `supportedOptions` has a `*`. Is that still possible? The usage of `*` would be the easiest path forward, but that might not be possible.

If not, the ones I can think of are presently coming from the AWS catalogs (though we should also look into the JDBC catalog as well).

**Additional Catalog Properties To Consider**:
_General Catalog Properties_
- `catalog-impl`
- `io-impl`

[_GlueCatalog Specific Options_](https://iceberg.apache.org/aws/#glue-catalog)
- `lock-impl`
- `lock.table`

[_DynamoDbCatalog Specific Options_](https://iceberg.apache.org/aws/#dynamodb-catalog)
- `dynamodb.table-name`

[_AWS Client Configuration_](https://iceberg.apache.org/aws/#aws-client-customization)
- `client.factory`

There are further ones which I'll try to link to in a bit (such as S3FileIO options for AWS authentication and encryption parameters), but those are the first ones I encountered in the docs.

Reply:
Also, will users be able to specify their own catalog implementation (with its own options) still?